woocommerce/woocommerce
# Monorepo Utils: Add Code Freeze date function

### Submission Review Guidelines:

- I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
- I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/). 
- Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:



Add a function to verify Code Freeze date and make it available to the CLI.

This PR introduces the idea of a `--github` flag to persist variables between Github Action runs. This idea is used extensively in Code Analyzer.



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

#### Test the CLI

1. `pnpm install`
2. Check if today is code freeze day, `pnpm utils code-freeze verify-day`
3. Check if a specific day is code freeze. Note that April 17th is indeed code freeze day but depending on where you are in the world messing around with the time (below is 3pm) should produce mixed results because calculations are based on UTC time. So play around and make sure it works as expected. Check the logging to make sure the logic makes sense.
```
pnpm utils code-freeze verify-day -o '2023-04-17T15:00:00.000'
```
4. Try passing in an invalid string. It should throw an error.
```
pnpm utils code-freeze verify-day -o 'invalid date string'
```

#### Test the Github Action

1. Push this branch to your fork and enable actions.
2. If today is not the day of code freeze check that the `Release: Code freeze` action runs and does not proceed after the first step of checking the day. Be sure to select this branch, `add/code-freeze-date` when running tests.
<img width="475" alt="image" src="https://user-images.githubusercontent.com/1922453/231364471-1836ca16-9174-4338-8f07-685f0e245aad.png">
 3. Now test the action should proceed on code freeze day by supplying the time override with a date string you know represents code freeze day, like `2023-04-17T15:00:00.000`. You can check `Skip Slack Ping` as well. The action should proceed to the next step. 
<img width="386" alt="image" src="https://user-images.githubusercontent.com/1922453/231366175-7d7bff40-f265-48fc-9cfd-9935c7115843.png">
